daemon/cmd/ipcache.go

```diff
@@ -102,91 +101,176 @@ func containsSubnet(outer, inner net.IPNet) bool {
-			Resource: restoredCIDRResource,
-			Metadata: []ipcache.IPMetadata{ipcachetypes.RequestedIdentity(nid)},
-		})
+// restoreIPCache recreated ipcache metadata entries from the dumped allocator and
+// bpf map state.
+//
+// The goal of this logic is to ensure, as much as possible, that local identities (i.e. CIDR)
+// get the same numeric identity upon agent restart.
+//
+// This reconstructs the ipcache state from the previous BPF map and the restored local
+// identities. Specifically, if a prefix in the ipcache has a CIDR label, this re-creates
+// that metadata entry.
+//
+// For ingress IPs, it will add those to the ipcache and configure the local node
+// accordingly.
+func (d *Daemon) restoreIPCache(localPrefixes map[netip.Prefix]identity.NumericIdentity, restoredIdentities map[identity.NumericIdentity]*identity.Identity) {
+	if len(localPrefixes) == 0 {
+		return
 	}
-	for _, prefix := range ingressIPs {
-		metaUpdates = append(metaUpdates, ipcache.MU{
-			Prefix:   prefix,
-			Source:   source.Restored,
-			Resource: ingressResource,
-			Metadata: []ipcache.IPMetadata{labels.LabelIngress},
-		})
 
-		// Set any restored ingress IPs back on the LocalNode object
-		d.nodeLocalStore.Update(func(n *node.LocalNode) {
-			addr := prefix.Addr()
-			if addr.Is4() {
-				n.IPv4IngressIP = addr.AsSlice()
-			} else {
-				n.IPv6IngressIP = addr.AsSlice()
+	metaUpdates := make([]ipcache.MU, 0, len(localPrefixes))
+	d.restoredCIDRs = make(map[netip.Prefix]identity.NumericIdentity, len(localPrefixes))
+	nidsToWithhold := []identity.NumericIdentity{}
+
+	for prefix, nid := range localPrefixes {
+		// Restore Ingress IPs as necessary
+		if nid == identity.ReservedIdentityIngress {
+			metaUpdates = append(metaUpdates, ipcache.MU{
+				Prefix:   prefix,
+				Source:   source.Restored,
+				Resource: ingressResource,
+				Metadata: []ipcache.IPMetadata{labels.LabelIngress},
+			})
+
+			// Set any restored ingress IPs back on the LocalNode object
+			d.nodeLocalStore.Update(func(n *node.LocalNode) {
+				addr := prefix.Addr()
+				if addr.Is4() {
+					n.IPv4IngressIP = addr.AsSlice()
+				} else {
+					n.IPv6IngressIP = addr.AsSlice()
+				}
+			})
+			log.WithField(logfields.Ingress, prefix).Info("Restored ingress IP")
+			continue
+		}
+
+		// For every prefix -> nid pair, look to see if there is a restored identity for this nid.
+		// If not, then request the same numeric identity *and* possibly insert CIDR labels
+		// If yes, **and** the identity contains the prefix `cidr:` label,
+		//   then upsert that exact set of labels in the ipcache.
+		id := restoredIdentities[nid]
+		if id == nil {
+			metadata := []ipcache.IPMetadata{ipcachetypes.RequestedIdentity(nid)}
+
+			// Check to see if any other prefixes share this identity. If not, then
+			// add CIDR labels as well.
+			found := false
+			for tryPrefix, tryNID := range localPrefixes {
+				if tryNID == nid && tryPrefix != prefix {
+					found = true
+					break
+				}
 			}
```

Comment:
Nit: I'm assuming you are aware of this, but this algorithmically quadratic in terms of prefixes, as `id == nil` will be true for every prefix during upgrades.

Assuming we're restoring a worst case of 1000 restored prefixes I'd expect this to take ~20 milliseconds (1M iterations, based on my recent map-of-prefixes benchmark) upon upgrade (but only upgrade, not regular agent restart). Probably in the range of acceptable. 

However, I do wonder if it makes sense to assemble a quick `counter.Counter[netip.prefix]` counting the identities per prefix and check that one here. In any case, not a blocker for me.